### PR TITLE
Get download link for documents in S3

### PR DESCRIPTION
**Pull request by bella-cockrell**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/DOC-613

### *What is the problem we're trying to solve*

This PR implements the download path for files in S3.  This is part of the file size feature that looks to increase the size of files in S3 

### *What changes have we introduced*

The endpoint now GETs a download link for the file in S3. The system does this by first searching for the claim via the url param. It then returns the S3 download link based on the document entity that is the child of the claim entity. We decided that this would be the best way, because a user should only have access to documents that they have a claim to. 

Tests also included!

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

In the future, the relationship between claims and documents may change, but we predict that many claims will have a relationship with one document (*:1), so this behaviour should not be affected. No follow-up actions!

### CODE REVIEW

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                result = _s3Gateway.GeneratePreSignedDownloadUrl(claim.Document);
+            }
+            catch (AmazonS3Exception e)
+            {
+                Console.WriteLine($"Error when retrieving the presigned URL: {e.Message}");
+                throw e;
+            }
+
+            return result;
```
Comment (bogdan-zaharia) [inline DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:42]: Suggestion you could move the return inside the try block and also remove line 30 and move "var" inside the try block before result

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+using System;
+using DocumentsApi.V1.Boundary.Response.Exceptions;
+using DocumentsApi.V1.Gateways.Interfaces;
+using DocumentsApi.V1.UseCase.Interfaces;
+using Amazon.S3;
+using System.Diagnostics.CodeAnalysis;
+
+namespace DocumentsApi.V1.UseCase
+{
+    public class DownloadDocumentWithUrlUseCase : IDownloadDocumentWithUrlUseCase
```
Comment (bogdan-zaharia) [inline DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:10]: We should try to make the name of this use case to reflect what it is doing.  This use case is not actually downloading the document, instead it creates a download link for the document which will then be used to access the document. Suggestion: "CreatePreSignedDownloadUrlUseCase" (or anything that makes sense)

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            if (claim == null)
+            {
+                throw new NotFoundException($"Cannot find a claim with ID: {claimId}");
+            }
+
+            var result = "";
+
+            try
+            {
+                result = _s3Gateway.GeneratePreSignedDownloadUrl(claim.Document);
```
Comment (bogdan-zaharia) [inline DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:34]: Does this function need to be awaited? (does the gateway function need to be async?)

**Review by bogdan-zaharia — COMMENTED** [review]
Looks good, left a few comments!

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs (hunk 1 above)
Comment (bella-cockrell) [reply DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:42]: Good point!

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs (hunk 3 above)
Comment (bella-cockrell) [reply DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:34]: Good point, but I don't think so -- I had a look at the AWS documentation and the function used in the gateway to call S3 appears to be synchronous in the [examples](https://docs.aws.amazon.com/sdkfornet1/latest/apidocs/html/M_Amazon_S3_AmazonS3_GetPreSignedURL.htm), so we don't need to await it anywhere else.

Additionally, in the [AWS forum](https://forums.aws.amazon.com/thread.jspa?threadID=171406) [login required], there is an AWS rep saying in 2015 that there is no async version of GetPreSignedUrl as it 'This method does not make any service calls, as it simply constructs a string'.

File: DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs (hunk 2 above)
Comment (bella-cockrell) [reply DocumentsApi/V1/UseCase/DownloadDocumentWithUrlUseCase.cs:10]: Another good point! I think in this instance I was following the naming convention for DownloadDocumentUseCase. I'll change the name of the classes!

**Review by bogdan-zaharia — APPROVED** [review]
LGTM

**Pull request merged**